benchmark/benchmark.hpp

```diff
@@ -441,20 +447,141 @@ class multiple_malloc_free_benchmark : public benchmark_interface<Size, Alloc> {
         }
     }
 
-    void prepareWorkload(benchmark::State &state) {
+    virtual void prepareWorkload(benchmark::State &state) {
         auto tid = state.thread_index();
         auto &n = next[tid];
+        using distribution = std::uniform_int_distribution<size_t>;
         std::default_random_engine generator;
-        distribution dist;
+        distribution dist_offset(0, max_allocs - 1);
+        distribution dist_opt_type(0, 1);
         generator.seed(0);
-        dist.param(distribution::param_type(0, max_allocs - 1));
         auto sizeGenerator = base::alloc_sizes[tid];
+        std::vector<size_t> free;
+        std::vector<size_t> allocated;
+
+        // this benchmark prealloc memory, so we start with some allocation
+        size_t i = 0;
+        for (; i < max_allocs / 2; i++) {
+            allocated.push_back(i);
+        }
+        for (; i < max_allocs; i++) {
+            free.push_back(i);
+        }
 
         n.clear();
         for (int64_t j = 0; j < state.max_iterations * allocsPerIterations;
              j++) {
-            n.push_back({dist(generator), sizeGenerator.nextSize()});
+            if (allocated.empty() ||
+                (dist_opt_type(generator) == 0 && !free.empty())) {
+
+                std::swap(free[dist_offset(generator) % free.size()],
+                          free.back());
+                auto offset = free.back();
+                free.pop_back();
+
+                n.push_back({true, offset, sizeGenerator.nextSize()});
+                allocated.push_back(offset);
+            } else {
+                std::swap(allocated[dist_offset(generator) % allocated.size()],
+                          allocated.back());
+                auto offset = allocated.back();
+                allocated.pop_back();
+
+                n.push_back({false, offset, 0});
+                free.push_back(offset);
+            }
         }
+
         next_iter[tid] = std::make_unique<next_alloc_data_iterator>(n.cbegin());
     }
 };
+
+// This class benchmarks performance randomly allocates and frees,
```

Comment:
improve grammar e.g. "// This class benchmarks performance by randomly allocating and freeing memory. // Initially, it slowly increases the memory footprint, and later decreases it."

Reply:
Done